### PR TITLE
[RFC] MSBuild\MMP Test Template Rework

### PR DESCRIPTION
**Pull request by chamons**
I decided to burn a day and prototype how I'd rewrite ProjectTestHelpers in a less organic\terrible way. My goals were:

- Not tied to XM concepts, so we could share it with XI.
- Shim into ProjectTestHelpers API for now, we can bite off reworking the tests later
   - This means there is some ugly conversion code that will go away in time
- Simple API, as we're just really just doing a fancy copy + sed
- Support a subset of what MMP\msbuild-mac does today. If people 👍 then easy to extend:
        - Project Flavor: Full, Modern, NetStandard, etc
        - Project Type: App, Lib, Binding Project
        - Language: C#, F#

This is what I came up with.

The PR line diff count is large because I also refactored out RunCommand from Driver.cs so we could stop duplicating it in test code. It may be useful to review the commits separately because of this.

The test template API works like this:

- If you wants custom template files, create a TemplateInfo describing what you need
- You instance a TemplateEngine subclass either with a TemplateInfo or your flavor\language
    - MacAppTemplateEngine
    - MacBindingTemplateEngine
    - MacLibraryTemplateEngine
    - NetStandardTemplateEngine
- Create a number of Substitutions classes (FileSubstitutions, PListSubstitutions, ProjectSubstitutions) which describe how you want to fill in the "holes" in the templates
- You call a Generate method passing in those substitutions and any additional options.

If we want to move forward with this, my plan would go as follows:

- Flesh out test-template with new components that contain the rest of the functionality ProjectTestHelpers has (Building projects, Frameworks, etc)
- Create an initial PR that makes ProjectTestHelpers use the new API and land it
- Create a second PR that updates the mmp/msbuild-mac tests to use the new API directly and remove the shim layer
- Create a third PR that moves\renames the template files to be in a more consistent location, in prep for iOS additional files
- Create one last PR that checked in one iOS template, the infrastructure to spin it up, and a test using it.

### CODE REVIEW

**Comment by chamons** [comment]
Apologies @spouliot, I did not add sufficient context. 

So you can consider this RFC a rewrite of this [file](https://github.com/xamarin/xamarin-macios/blob/master/tests/common/mac/ProjectTestHelpers.cs) into parts that are more manageable and not hard coded to XM.

ProjectTestHelpers is the templating system XM uses for mmp and msbuild tests, as described in the paper document going around. It has a number of problems, the worst two being:

- Very organic growth makes knowing which config options are supported by which APIs very difficult
- Very tied to XM concepts, so not easy to share with XI

Please see the last section in the paper document for a comparison with the XA approach.

I have a few more commits I've done today that break out a few more pieces (Classic, System Mono). The full plan in listed above, but roughly we're done when I can `git rm tests/common/mac/ProjectTestHelpers.cs`

File: tools/test-template/test-template-mac/Generator/TemplateInfo.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+							return "FSharpUnifiedExample.fsproj";
+						case ProjectType.Library:
+							return "FSharpUnifiedLibrary.fsproj";
+						}
+						break;
+					}
+					break;
+				}
+
+				throw new NotImplementedException ();
```
Comment (rolfbjarne) [inline tools/test-template/test-template-mac/Generator/TemplateInfo.cs]:
```suggestion
				throw new NotImplementedException ($"Language: ${Language} Flavor: ${Flavor}");
```

File: tools/test-template/test-template-mac/Utilities/DirectoryFinder.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+﻿using System;
+using System.IO;
+using System.Reflection;
+
+namespace Xamarin.Tests.Templating
+{
+	static class DirectoryFinder
+	{
+		public static string FindRootDirectory ()
```
Comment (rolfbjarne) [inline tools/test-template/test-template-mac/Utilities/DirectoryFinder.cs]: The Configuration class already has API that does the same thing:

https://github.com/xamarin/xamarin-macios/blob/f389c03970b0b48b3300340d2cd54a0af6f44fa8/tests/common/Configuration.cs#L329-L333

File: tools/test-template/test-template-mac/Utilities/DirectoryFinder.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+		public static string FindTestDirectory => Path.Combine (FindRootDirectory (), "..", "tests") + "/";
+
+		public static string FindSourceDirectory ()
+		{
+			string codeBase = Assembly.GetExecutingAssembly ().CodeBase;
+			UriBuilder uri = new UriBuilder (codeBase);
+			string path = Uri.UnescapeDataString (uri.Path);
+			string assemblyDirectory = Path.GetDirectoryName (path);
+			return Path.Combine (assemblyDirectory, FindTestDirectory + "common/mac");
```
Comment (rolfbjarne) [inline tools/test-template/test-template-mac/Utilities/DirectoryFinder.cs]: Doesn't this end up with something like `/build/dir/xamarin-macios/tests/something/something/build/dir/tests/common/mac`?

Seems like you just want:

```suggestion
			return Path.Combine (FindTestDirectory, "common", "mac");
```

**Comment by chamons** [comment]
I'm going to close this and open RFC issue shortly.

**Pull request closed without merging**